eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2021 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.io
+
+import akka.actor.{ActorContext, ActorRef}
+import akka.testkit.{TestFSMRef, TestProbe}
+import fr.acinq.bitcoin.Crypto.PublicKey
+import fr.acinq.eclair.FeatureSupport.Optional
+import fr.acinq.eclair.TestConstants.{Alice, Bob}
+import fr.acinq.eclair.blockchain.DummyOnChainWallet
+import fr.acinq.eclair.io.Peer.ChannelFactory
+import fr.acinq.eclair.message.OnionMessages
+import fr.acinq.eclair.message.OnionMessages.{IntermediateNode, Recipient}
+import fr.acinq.eclair.wire.protocol
+import fr.acinq.eclair.wire.protocol.{NodeAddress, OnionMessage}
+import fr.acinq.eclair.{Features, NodeParams, TestKitBaseClass, randomKey}
+import org.scalatest.Outcome
+import org.scalatest.funsuite.FixtureAnyFunSuiteLike
+
+case class FakeChannelFactory(channel: TestProbe) extends ChannelFactory {
+  override def spawn(context: ActorContext, remoteNodeId: PublicKey, origin_opt: Option[ActorRef]): ActorRef = {
+    channel.ref
+  }
+}
+
+class MessageRelaySpec extends TestKitBaseClass with FixtureAnyFunSuiteLike {
```

Comment:
If you follow my suggestion to have the switchboard spawn the message relay actors, I think this test suite can be greatly simplified. It should simply spawn message relay actors with test probes in all the `ActorRef` parameters, send the messages the `MessageRelay` actor expects and check its responses on the test probes.

Separately of that unit test suite, you should add end to end tests in `PaymentIntegrationSpec` to verify that the complete flow of sending an onion message works.

Reply:
Done